### ISSUE TITLE
Review _EqualityOperator

### ISSUE THREAD

**Issue by mmatera** [mathics/Mathics#1197]
Looking at #1169 and the question about loading the KnotTheory package, was looking at the exception raised on a comparison. I tracked that up to the class `_EqualityOperator`, where I found several things that I do not understand in the flow.

* If we call a==b==c==d, the code in master checks all the pairs (a==a, b==b, c==c, d==, a==b, a==c, a==d, b==c,b==d, c==d). However, it would be enough to check a=b, b==c, c==d and a==d.
* if a and b are two strings, the result of the comparison seems to be `False`
* The class has two `apply*` methods: one for inexact comparisons, and another general. In the general method (`apply`), the inexact comparison is called if all the elements to compare are exact (??? )

**Comment by rocky** [comment]
@mmatera If you come up with test cases that shows the bugs, I'll look at when I get a chance.

**Comment by soehms** [comment]
Thanks for investigating this!

**Comment by mmatera** [comment]
The discovering of this issue comes from the line 
```
Knot[S_String?(StringMatchQ[#,StringExpression[DigitCharacter.., "_"|" ", DigitCharacter..]]&)]/;((#[[1]]<=10\[And]#[[2]]<=NumberOfKnots[#[[1]]])&[ToExpression/@StringSplit[S,"_"|" "]]):=Knot@@(ToExpression/@StringSplit[S,"_"|" "])
```
that rises an exception

**Comment by mmatera** [comment]
Mostly addressed by #1209
